tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/passes/ConfigurationPassImpl.scala

```diff
@@ -249,32 +305,96 @@ class ConfigurationPassImpl @Inject()(val options: WriteablePassOptions,
+      )
+      val msg =
+        s"$contextName: Expected $specName to be in the canonical form Init /\\ [][Next]_vars /\\ ..."
+      throw new ConfigurationError(msg)
+    }
+
     module.operDeclarations.find(_.name == specName) match {
       case None =>
-        throw new ConfigurationError(s"$contextName: Operator $specName not found (used as SPECIFICATION)")
-
-        // the canonical form: Init /\ [][Next]_vars /\ ...
-      case Some(TlaOperDecl(_, List(),
-        OperEx(TlaBoolOper.and,
-          // Init
-          OperEx(TlaOper.apply, NameEx(init)),
-          // [][Next]_vars
-          OperEx(TlaTempOper.box,
-            // [Next]_vars
-            OperEx(TlaActionOper.stutter,
-              // Next
-              OperEx(TlaOper.apply, NameEx(next)),
-              // vars
-            _*
-          )), ///
-          _*))) =>
-        (init, next)
+        throw new ConfigurationError(
+          s"$contextName: Operator $specName not found (used as SPECIFICATION)"
+        )
+
+      // the canonical form: Init /\ [][Next]_vars /\ ...
+      case Some(decl @ TlaOperDecl(_, List(), spec)) =>
+        flattenAnds(spec) match {
+          case OperEx(TlaBoolOper.and, args @ _*) =>
+            val (nonFairness, fairness) = args.partition {
+              case OperEx(TlaTempOper.weakFairness, _*)   => false
+              case OperEx(TlaTempOper.strongFairness, _*) => false
+              case _                                      => true
+            }
+            if (nonFairness.length != 2) {
+              throwError(decl)
+            } else {
+              if (fairness.nonEmpty) {
+                val es = fairness.mkString(" /\\ ")
+                val msg = s"Fairness constraints are ignored by Apalache: $es"
+                logger.warn(msg)
+              }
+              nonFairness match {
+                case Seq(
+                      // Init
+                      OperEx(TlaOper.apply, NameEx(init)),
+                      // [][Next]_vars
+                      OperEx(
+                        TlaTempOper.box,
+                        OperEx(
+                          TlaActionOper.stutter,
+                          OperEx(TlaOper.apply, NameEx(next)),
+                          _*
+                        )
+                      )
+                    ) =>
+                  (init, next)
+
+                case Seq(
+                      // [][Next]_vars
+                      OperEx(
+                        TlaTempOper.box,
+                        OperEx(
+                          TlaActionOper.stutter,
+                          OperEx(TlaOper.apply, NameEx(next)),
+                          _*
+                        )
+                      ),
+                      // Init
+                      OperEx(TlaOper.apply, NameEx(init))
+                    )
+                    =>
+                  (init, next)
+              }
+            }
```

Comment:
The logic of the control flow here seems a bit convoluted, and it will add to our partial match warnings. What about doing something like this instead?

```suggestion
nonFairness match {
  case Seq(
    OperEx(TlaOper.apply, NameEx(init)), // Init
    OperEx(TlaTempOper.box, OperEx(TlaActionOper.stutter, OperEx(TlaOper.apply, NameEx(next)), _*)) // [][Next]_vars
  ) |
      Seq(
        OperEx(TlaTempOper.box, OperEx(TlaActionOper.stutter, OperEx(TlaOper.apply, NameEx(next)), _*)),
        OperEx(TlaOper.apply, NameEx(init)) // Init
      ) =>
    if (fairness.nonEmpty) {
      val es = fairness.mkString(" /\\ ")
      val msg = s"Fairness constraints are ignored by Apalache: $es"
      logger.warn(msg)
    }
    (init, next)
  case _ => throwError(decl)
}
```

Reply:
Oh yes, this looks much better. I was also struggling with all these cases in pattern matching.